src/context.rs

```diff
@@ -33,76 +34,91 @@
-    pub batches:
-        HashMap<BatchNumber, (BatchCount, Vec<serde_json::value::Value>, ResponsesCallback)>,
+/// Groups parameters for each request.
+pub enum RequestParams<T> {
+    /// Replicates the same list of parameters for all language servers in a context.
+    All(Vec<T>),
+    /// Uses different parameters for each language server in a context.
+    Each(HashMap<ServerName, Vec<T>>),
+}
+
+pub struct ServerSettings {
+    pub root_path: String,
+    pub offset_encoding: OffsetEncoding,
+    pub preferred_offset_encoding: Option<OffsetEncoding>,
     pub capabilities: Option<ServerCapabilities>,
-    pub completion_items: Vec<CompletionItem>,
+    pub tx: Sender<ServerMessage>,
+}
+
+pub struct Context {
+    batch_count: BatchCount,
+    pub batch_sizes: HashMap<BatchNumber, HashMap<ServerName, usize>>,
+    pub batches: HashMap<
+        BatchNumber,
+        (
+            Vec<(ServerName, serde_json::value::Value)>,
+            ResponsesCallback,
+        ),
+    >,
+    pub code_lenses: HashMap<String, Vec<(ServerName, CodeLens)>>,
+    pub completion_items: Vec<(ServerName, CompletionItem)>,
     pub completion_items_timestamp: i32,
     // We currently only track one client's completion items, to simplify cleanup (else we
     // might need to hook into ClientClose). Track the client name, so we can check if the
     // completions are valid.
     pub completion_last_client: Option<String>,
     pub config: Config,
+    pub diagnostics: HashMap<String, Vec<(ServerName, Diagnostic)>>,
+    pub documents: HashMap<String, Document>,
     pub dynamic_config: DynamicConfig,
-    pub diagnostics: HashMap<String, Vec<Diagnostic>>,
-    pub code_lenses: HashMap<String, Vec<CodeLens>>,
     pub editor_tx: Sender<EditorResponse>,
-    pub lang_srv_tx: Sender<ServerMessage>,
-    pub language_id: String,
-    pub outstanding_requests: HashMap<(&'static str, String, Option<String>), OutstandingRequests>,
+    pub language_id: LanguageId,
+    pub language_servers: BTreeMap<ServerName, ServerSettings>,
+    pub outstanding_requests:
+        HashMap<(ServerName, &'static str, String, Option<String>), OutstandingRequests>,
     pub pending_requests: Vec<EditorRequest>,
     pub pending_message_requests: VecDeque<(Id, ShowMessageRequestParams)>,
     pub request_counter: u64,
     pub response_waitlist: HashMap<Id, (EditorMeta, &'static str, BatchNumber, bool)>,
-    pub root_path: String,
     pub session: SessionId,
-    pub documents: HashMap<String, Document>,
-    pub offset_encoding: OffsetEncoding,
-    pub preferred_offset_encoding: Option<OffsetEncoding>,
     pub work_done_progress: HashMap<NumberOrString, Option<WorkDoneProgressBegin>>,
     pub work_done_progress_report_timestamp: time::Instant,
-    pub pending_file_watchers: HashMap<Option<PathBuf>, Vec<CompiledFileSystemWatcher>>,
+    pub pending_file_watchers:
+        HashMap<(ServerName, String, Option<PathBuf>), Vec<CompiledFileSystemWatcher>>,
```

Comment:
The "To server" and "From server" logs need to be adapted. Right now I can't tell which server the duplicate requests are directed at.

I noticed this when I tried

	[language.foo.c_cpp]
	filetypes = ["c", "cpp"]
	roots = ["compile_commands.json", ".clangd", ".git", ".hg"]
	command = "clangd"
	[language.foo.protobuf]
	filetypes = ["cpp"]
	roots = ["compile_commands.json", ".clangd", ".git", ".hg"]
	command = "pls"

trying to get go-to-definition for protobuf messages inside C++.
Sadly it doesn't work, clangd always responds before [pls](https://github.com/lasorda/protobuf-language-server/)

I think this one we could solve by ranking `pls` before `clangd`.
Defining `pls` as "primary" server feels odd but would work here.

Reply:
Can kak-lsp currently handle gotos for multiple results? I know tsserver does return multiple results at some times. If so, we can go with merging the results instead of picking the first non-empty one. I think ranking them is not the ideal solution, because we would omit results anyway. What do you think?

Reply:
Btw, I just remembered, the way I set up this was to use a `BTreeMap` for servers, therefore currently they are lexicographically ordered, but it should be simple to add a custom ordering logic that respects a ranking value.

As for gotos, it's possible to merge results, but on Kakoune side, it uses a single root. I think this root should come from the "priority" server.

Reply:
> Can kak-lsp currently handle gotos for multiple results?

yeah in this case it will call `lsp-show-goto-choices` which shows a location list buffer.
We could maybe make that call lsp-menu instead

> If so, we can go with merging the results instead of picking the first non-empty one. I think ranking them is not the ideal solution, because we would omit results anyway. What do you think?

AFAICT the current behavior is the "first non-empty one" meaning it's a race between all language servers.
It would be good to make this deterministic. The simplest approach would be to wait for all responses (perhaps with a timeout) and then combine them. If that's too much work it can be a follow-up but let's keep that in mind.

> Btw, I just remembered, the way I set up this was to use a BTreeMap for servers, therefore currently they are lexicographically ordered, but it should be simple to add a custom ordering logic that respects a ranking value.

Good. We have two possible behaviors, either give the highest-ranked non-empty response, or merge all responses.
For my use case I'd like the first. Maybe there are uses for the second.

Reply:
I pushed a commit that merges the results. For goto specifically, I think it's better UX to present all possible results. The logic for choosing the highest-ranked would be hard to mix with the merging strategy.

Reply:
> For goto specifically, I think it's better UX to present all possible results.

yes let's show all of them for now, but remove duplicates